### PR TITLE
Release 0.1.0

### PR DESCRIPTION
**Pull request by kevinkirkup**

### CODE REVIEW

File: pureport/resource_aws_connection.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+func resourceAWSConnection() *schema.Resource {
+
+	connection_schema := map[string]*schema.Schema{
+		"cloud_service_hrefs": {
+			Type:     schema.TypeList,
+			Optional: true,
+			Elem:     &schema.Schema{Type: schema.TypeString},
+		},
+		"peering": {
```
Comment (dmittendorf) [inline pureport/resource_aws_connection.go]: We should name this `peering_type` for consistency with the Ansible module https://github.com/pureport/pureport-ansible-modules/blob/develop/ansible/modules/pureport/pureport_aws_direct_connect_connection.py

File: pureport/resource_azure_connection.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+func resourceAzureConnection() *schema.Resource {
+
+	connection_schema := map[string]*schema.Schema{
+		"service_key": {
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+		"peering": {
```
Comment (dmittendorf) [inline pureport/resource_azure_connection.go]: Please rename to `peering_type` for consistency with Ansible modules.

File: pureport/resource_google_cloud_connection.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	if description, ok := d.GetOk("description"); ok {
+		c.Description = description.(string)
+	}
+
+	if highAvailability, ok := d.GetOk("high_availability"); ok {
+		c.HighAvailability = highAvailability.(bool)
+	}
+
+	// Google Optionals
+	if secondaryPairingKey, ok := d.GetOk("secondary_pairing_key"); ok {
```
Comment (dmittendorf) [inline pureport/resource_google_cloud_connection.go]: Is there a way to require `secondaryPairingKey` when highAvailability is true?

File: pureport/resource_network.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"account_id": {
```
Comment (dmittendorf) [inline pureport/resource_network.go]: `account_href`?

File: pureport/connection.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+			Type:     schema.TypeBool,
+			Optional: true,
+			ForceNew: true,
+		},
+		"location_href": {
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+		"network": {
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/connection.go]: Currently for networks, we need both the `id` and the `href` so kept it as a link object.

File: pureport/connection.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+	if data, ok := d.GetOk("customer_networks"); ok {
+
+		customerNetworks := []client.CustomerNetwork{}
+
+		for _, cn := range data.([]map[string]string) {
+
+			new := client.CustomerNetwork{
+				Name:    cn["name"],
+				Address: cn["Address"],
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/connection.go]: yes

File: README.md [hunk 7]
```diff
@@ -1,2 +1,89 @@
...
+- Website: https://www.terraform.io
+- [![Gitter chat](https://badges.gitter.im/hashicorp-terraform/Lobby.png)](https://gitter.im/hashicorp-terraform/Lobby)
+- Mailing list: [Google Groups](http://groups.google.com/group/terraform-tool)
+
+<img src="https://cdn.rawgit.com/hashicorp/terraform-website/master/content/source/assets/images/logo-hashicorp.svg" width="600px">
+
+Maintainers
+-----------
+
+This provider plugin is maintained by the Terraform team at [HashiCorp](https://www.hashicorp.com/).
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply README.md]: Yeah, this was just a cut/paste from one of the existing Terraform roles.

File: pureport/data_source_locations.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+						},
+						"href": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"links": {
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_locations.go]: This is actually the `LocationLinkConnection` not the `Link` type from Kato, so I think we are OK here.
Agree on the remove of Link from this interface though.

File: pureport/data_source_cloud_services.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+						},
+						"ipv4_prefix_count": {
+							Type:     schema.TypeInt,
+							Computed: true,
+						},
+						"ipv6_prefix_count": {
+							Type:     schema.TypeInt,
+							Computed: true,
+						},
+						"cloud_region_id": {
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_cloud_services.go]: Currently in the API, this is referencing the `ID` of the cloud region that the service is located, not the `href` so it this case, I think it's OK.

i.e. 
1	*aws-s3-us-west-1*	AWS S3 us-west-1	AWS	S3	4	FALSE		1557431944

File: pureport/data_source_networks.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"account_id": {
```
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_networks.go]: Same here. The API is referencing the `ID`

ie. 
```
func (a *NetworksApiService) FindNetworks(ctx context.Context, accountId string) ([]Network, *http.Response, error)
```

File: pureport/resource_aws_connection.go (hunk 1 above)
Comment (kevinkirkup) [reply pureport/resource_aws_connection.go]: ok

File: pureport/resource_azure_connection.go (hunk 2 above)
Comment (kevinkirkup) [reply pureport/resource_azure_connection.go]: ok

File: pureport/resource_network.go (hunk 4 above)
Comment (kevinkirkup) [reply pureport/resource_network.go]: Same as above. The Account ID is required by the API.

```
func (a *NetworksApiService) AddNetwork(ctx context.Context, accountId string, localVarOptionals *AddNetworkOpts) (*http.Response, error)
```

File: pureport/resource_google_cloud_connection.go (hunk 3 above)
Comment (kevinkirkup) [reply pureport/resource_google_cloud_connection.go]: Let me check.

File: pureport/data_source_cloud_services.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_cloud_services.go]: We can revisit this. We may be able to get rid of this completely.
Not sure that it's required.

[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_cloud_services.go]: Or abstract, the interface to only expose the provider id.
Need to think about that a bit more.

[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_cloud_services.go]: I'm thinking in general what we need to do here is provide both the `id` and `href` parameters for reading but only require `href` for writing `Link` objects.
There are a couple of reasons for this:
* There are a few of places in the golang SDK that require providing an ID (`accountID` for creating networks, and `networkID` for creating connections.
* It's easier to centralize the HREF parser for extracting the ID in a single location. i.e. Kato and have just be able to provide this to clients.

Thoughts?

[the remark this answers is not in the document]
Comment (kevinkirkup) [reply pureport/data_source_cloud_services.go]: Moving to use `HREF` everywhere.

File: pureport/resource_google_cloud_connection.go (hunk 3 above)
Comment (kevinkirkup) [reply pureport/resource_google_cloud_connection.go]: Doesn't look like there is, but it should be validated by the server so we should be ok.

File: go.mod [hunk 11]
```diff
@@ -0,0 +1,13 @@
+module github.com/pureport/terraform-provider-pureport
+
+go 1.12
+
+require (
+	github.com/antihax/optional v0.0.0-20180407024304-ca021399b1a6
+	github.com/cenkalti/backoff v2.1.1+incompatible
+	github.com/hashicorp/terraform v0.11.13
+	github.com/pureport/pureport-sdk-go v0.0.0-20190506215644-d100251c07ec
```
Comment (Try75) [inline go.mod]: Does this need to be updated to a released version of pureport-sdk-go?

Comment (kevinkirkup) [reply go.mod]: Probably.

File: version/version.go [hunk 12]
```diff
@@ -0,0 +1,6 @@
+package version
+
+var (
+	// ProviderVersion is set during the release process to the release version of the binary
+	ProviderVersion = "dev"
```
Comment (Try75) [inline version/version.go]: Might need to change this for release, too.

Comment (kevinkirkup) [reply version/version.go]: So it looks like I took this from the Google Terraform Module and it looks like they just leave this as `dev`. I think this is to just define the `ProviderVersion` variable and default it to `dev`. During their build process, they specify the version on the build command line, which is what we will probably do as well.

File: go.mod (hunk 11 above)
Comment (kevinkirkup) [reply go.mod]: Should be fixed with the latest PR.

**Pull request merged**